main.go

```diff
@@ -233,69 +241,68 @@ func main() {
 		}, &routableLinks, &liveContainerIDs)
 
 		if err != nil {
-			fail(ctx, err)
+			doCleanup(err)
 		}
-		success(ctx)
+		doCleanup(nil)
 	}
 
-	success(ctx)
+	doCleanup(nil)
 }
 
-func startUnit(ctx context.Context, yml config.EazyYml, openPortsLocally bool) {
+func startUnit(ctx context.Context, yml config.EazyYml, openPortsLocally bool, runtime runtimes.ContainerRuntime) {
+	doCleanup := cleanUp(ctx, &runtime)
+
 	if len(yml.Integration.Bootstrap) > 0 {
-		_, err := utils.StartContainerByEazyYml(ctx, yml, config.GetLatestIntegrationImageName(yml), config.DockerConfig{
+		_, err := runtime.StartContainerByEazyYml(ctx, yml, config.GetLatestIntegrationImageName(yml), config.RuntimeConfig{
 			Command:     yml.Integration.Bootstrap,
 			Wait:        true,
 			ExposePorts: false,
 		}, &routableLinks, &liveContainerIDs)
 
 		if err != nil {
-			fail(ctx, err)
+			doCleanup(err)
 		}
 	}
-	_, err := utils.StartContainerByEazyYml(ctx, yml, "", config.DockerConfig{
+	_, err := runtime.StartContainerByEazyYml(ctx, yml, "", config.RuntimeConfig{
 		Env:         yml.Deployment.Env,
 		Wait:        false,
 		ExposePorts: openPortsLocally,
 		IsRootImage: true,
 	}, &routableLinks, &liveContainerIDs)
 	if err != nil {
-		fail(ctx, err)
+		doCleanup(err)
 	}
 	if len(yml.Deployment.Health) > 0 {
-		_, err := utils.StartContainerByEazyYml(ctx, yml, config.GetLatestIntegrationImageName(yml), config.DockerConfig{
+		_, err := runtime.StartContainerByEazyYml(ctx, yml, config.GetLatestIntegrationImageName(yml), config.RuntimeConfig{
 			Command:     yml.Deployment.Health,
 			Wait:        true,
 			ExposePorts: false,
 		}, &routableLinks, &liveContainerIDs)
 		if err != nil {
-			fail(ctx, err)
+			doCleanup(err)
 		}
 	}
 }
 
-func success(ctx context.Context) {
-	cleanUp(ctx, 0, nil)
-}
-
-func fail(ctx context.Context, err error) {
-	cleanUp(ctx, 1, err)
-}
-
-func cleanUp(ctx context.Context, exitCode int, err error) {
-	log.Println("cleaning up running containers...")
-	term.RestoreTerminal(os.Stdout.Fd(), oldStateOut)
-	for _, id := range liveContainerIDs {
-		err := utils.KillContainer(ctx, id)
+func cleanUp(ctx context.Context, runtime *runtimes.ContainerRuntime) func(err error) {
```

Comment:
I'm not married to this if you don't care for the pattern, it just made refactoring to the interface less tedious.

Reply:
Yours is better :-)